### ISSUE TITLE
Create a bot to clean up forum posts after X days of age

### ISSUE THREAD

**Issue by kana-ki** [kana-ki/fidobot#1]
# Problem Space

Discord provides forum channels where users can post threads. ENE is using this for their users to post venue openings and work ads. They would like to keep all posts within the forums relevant, and clear them after X amount of days. "Open today" forums would be cleared after 24 hours, while "Open in a few days" forums would be cleared after 3 days, for example. 

Discord provides an auto-archive option for forums however this is unreliable and only simply removes a post from side navigation from followers after that time, it does not remove it from the forums thread index. 

# Suggested solution

Create a bot where on a periodic timer it checks for threads older than a per-forum-configurable age and deletes them.
We want to call it Fido because it's a dog that eats post :D

# Acceptance Criteria

- It can be enabled on a particular forum channel
- It removes posts older than X amount of days in that particular forum
- X is configurable by users with Manage Posts(?) permission
- X is configurable on a per forum channel basis
- It can be disable on a particular forum channel after
- Posts in a forum already over X days in age are immediately removed when it's enable for that channel (so users can see it working)

# Suggested design

- `/eat` : Configure the current (or specified) thread to be deleted. If used on a forum, configures default behaviour for existing and/or future threads.
  - Time Type : Choose between seconds, minutes, hours or days.
  - Time Value : Specify in how many Time Type.
  - Channel (Defaults to current if used in a thread) : Specify a thread or forum.
  - Eat Existing (Forums only) : Yes or no choice to decide whether Fido should eat existing threads according to specified time options.
  - Eat Future (Forums only) : Yes or no choice to decide whether Fido should be enabled for future threads created in this forum.
- `/donteat` : Disables Fido for the current (or specified) thread or forum. This command overrides forum configuration if fido is enabled on it.
  - Channel (Defaults to current if used in a thread) : Specify a thread of forum.

# Testing Considerations

## Happy path
- It can be enabled on a channel
- It can have a max post age set (in hours)
- It deletes posts older than set age
- It does not delete posts younger than set age
- It deletes posts older than set age even if there conversation within younger than that age
- It can be enabled on more than one forum
- It can have a max age set for each forum
- The max age set on one forum does not affect another's
- It deletes posts older than set age that were created before it was enable on the channel
- It can delete all kinds of posts (empty posts, posts where user not longer exists, posts with embed and images)
- It can delete locked posts
- It can delete archived posts
- The /eat command cannot be used by a new user (without Manage Posts permission)
- The /donteat command cannot be used by a new user (without Manage Posts permission)
- It works in multiple guilds (servers)
- I can set up the bot to work with multiple forums in multiple servers simultaneously
- I can change the maximum age after having first set it (re-execute the /eat command)

## Sad path
- What happens if max post age is set to 0
- What happens if someone uses /donteat on a channel it is already not eating
- What happens if enabled on a forum then that forum is deleted
- What happens if someone uses /eat or /donteat on non-forum channels

**Comment by kana-ki** [comment]
Looking good so far, clean, no over-engineering. Just enough to get it going - no need for much abstraction/infra.
I've updated the description to your design proposal- I like them, more flexible, and still simple for us and them.

**Comment by Exorion1er** [comment]
Thank you ! I've also just realized my readme looks scuffed lol, I'll fix that tomorrow 👍

**Comment by kana-ki** [comment]
As FYI for the ConfigurationBuilder, production will probably be Environment Variables as a config source. Has most flexibility for containerising and/or hosting in pretty much any cloud platform.

**Comment by Exorion1er** [comment]
Ah sounds good. Mind if I use preprocessor clauses to use them only on release builds ? It's just more practical with my workflow

**Comment by kana-ki** [comment]
You can chain the sources using ConfigurationBuilder, so it's like a fallback pattern. 
Priority is given to latest source given. 

```
var config = new ConfigurationBuilder()
                 .AddJsonFile("config.json", optional: true)
                 .AddUserSecrets<Propgram>(optional: true)
                 .AddEnvironmentVariables("FIDO_")
                 .Build();
```

**Comment by Exorion1er** [comment]
Once again, C# keeps blowing my mind. Thanks !

**Comment by Exorion1er** [comment]
Thread saving and eating after a fixed interval now works properly, with proper responses to user.
Next, I'm gonna work on handling forums and assigning their default behaviour.

**Comment by Exorion1er** [comment]
The `/eat` command logic is finished for both threads and forums.
I have 2 things to do next :
- The `/donteat` command which should be fairly simple
- Add permissions check to both commands, as right now, anyone could use them.

I also want to rewrite and clean up the bot responses, as it is pretty verbose right now.
Lastly, I also want to add a simple command to check current configurations on a server.

**Comment by Exorion1er** [comment]
Finished the `/donteat` command
Added permissions checks
Also added the `/sniff` command to list saved configurations

I'm gonna polish everything, do a bunch of tests to see if everything fits the spec sheet, rewrite some of the outputs to make it cleaner, then do a pull request 👍

**Comment by kana-ki** [comment]
With me on testing and code review!

**Comment by kana-ki** [comment]
Happy path

- [x] It can be enabled on a channel
- [x] It can have a max post age set (in hours)
- [x] It deletes posts older than set age
- [x] It does not delete posts younger than set age
- [x] It deletes posts older than set age even if there conversation within younger than that age
- [x] It can be enabled on more than one forum
- [x] It can have a max age set for each forum
- [x] The max age set on one forum does not affect another's
- [x] It deletes posts older than set age that were created before it was enable on the channel
- [x] It can delete all kinds of posts (empty posts, posts where user not longer exists, posts with embed and images)
- [x] It can delete locked posts
- [x] It can delete archived posts
- [x] The /eat command cannot be used by a new user (without Manage Posts permission)
- [x] The /donteat command cannot be used by a new user (without Manage Posts permission)
- [x] It works in multiple guilds (servers)
- [x] I can set up the bot to work with multiple forums in multiple servers simultaneously
- [x] I can change the maximum age after having first set it (re-execute the /eat command)

Sad path

- [x] What happens if max post age is set to 0
- [x] What happens if someone uses /donteat on a channel it is already not eating
- [x] What happens if enabled on a forum then that forum is deleted
- [x] What happens if someone uses /eat or /donteat on non-forum channels

**Comment by kana-ki** [comment]
Testing complete! ♥ 

**Issues found**
- [x] Executing any command in a channel (not thread) without specifying a forum name fails and does not give feedback as to what went wrong / what the user did wrong. 
- [x] The time shown on /sniff is in UTC but it's not indicating UTC
- [x] Closed/Archived posts are not deleting
- [x] fidotest command needs removing before mergies

**Nit picks**
- [x] "Will I eat existing threads ? Yes" When they read this, it's already done it, so maybe rephrase?
- [x] Sniff reveals what the EatExisting value was and when it was executed - but since it's not relevant after the initiate execution, to save confusion and make the /sniff lean maybe emit those?
  - [x] May want them if they're wondering why a thread wasn't deleted after setting `EatExisting` to `false`? Maybe only show if it false then.

**Comment by Exorion1er** [comment]
All done ! More testing and review pwease <3

**Comment by kana-ki** [comment]
Testing complete, code review has some pending changes. Kly will work on them on Monday.

**Comment by kana-ki** [comment]
Deployed and invite link delivered to Felaris
